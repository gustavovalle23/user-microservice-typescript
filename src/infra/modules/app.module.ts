import { Module } from '@nestjs/common';
import { UsersModule } from './user.module';
import { CommonModule } from './common.module';

@Module({
  imports: [CommonModule, UsersModule],
  controllers: [],
  providers: [],
})
export class AppModule {}
