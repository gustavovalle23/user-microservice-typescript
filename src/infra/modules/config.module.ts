import { Module } from '@nestjs/common';
import { ConfigModule as NestConfigModule } from '@nestjs/config';

@Module({
  imports: [
    NestConfigModule.forRoot({
      envFilePath: ['.env'],
      isGlobal: true,
    }),
  ],
})
export class ConfigModule {}
