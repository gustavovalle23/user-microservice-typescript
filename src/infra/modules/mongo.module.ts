import { Module } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { MongooseModule } from '@nestjs/mongoose';

@Module({
  imports: [
    MongooseModule.forRootAsync({
      useFactory: async (configService: ConfigService) => ({
        uri: configService.get<string>('MONGODB_URL'),
      }),
      inject: [ConfigService],
    }),
  ],
})
export class MongoModule {}
