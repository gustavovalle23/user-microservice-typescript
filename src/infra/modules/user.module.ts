import { Module } from '@nestjs/common';

import { UserResolver } from '@/infra/resolvers';
import { MongooseModule } from '@nestjs/mongoose';
import { User, UserSchema } from '@/infra/models';
import { UserRepository } from '@/infra/repositories';
import { JwtGateway } from '@/infra/gateways';
import { USER_REPOSITORY } from '@/domain/contracts/repo';
import { JWT_SERVICE } from '@/domain/contracts/gateways';
import { FindUserUseCase } from '@/use-cases/find';
import { FindAllUsersUseCase } from '@/use-cases/list';
import { CreateUserUseCase } from '@/use-cases/create';

@Module({
  imports: [
    MongooseModule.forFeature([
      {
        name: User.name,
        schema: UserSchema,
      },
    ]),
  ],
  providers: [
    UserResolver,
    FindUserUseCase.UseCase,
    FindAllUsersUseCase.UseCase,
    CreateUserUseCase.UseCase,

    {
      provide: USER_REPOSITORY,
      useClass: UserRepository,
    },
    {
      provide: JWT_SERVICE,
      useClass: JwtGateway,
    },
  ],
})
export class UsersModule {}
