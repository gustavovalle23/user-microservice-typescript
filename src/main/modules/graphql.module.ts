import { ApolloDriver } from '@nestjs/apollo';
import { Module } from '@nestjs/common';
import { GraphQLModule } from '@nestjs/graphql';
import { ApolloServerPluginLandingPageLocalDefault } from 'apollo-server-core';

@Module({
  imports: [
    GraphQLModule.forRoot({
      driver: ApolloDriver,
      autoSchemaFile: '../../schema.gql',
      debug: true,
      sortSchema: true,
      playground: false,
      plugins: [ApolloServerPluginLandingPageLocalDefault()],
    }),
  ],
})
export class GraphqlModule {}
