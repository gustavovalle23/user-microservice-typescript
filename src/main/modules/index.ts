export * from './app.module';
export * from './common.module';
export * from './graphql.module';
export * from './mongo.module';
export * from './user.module';
