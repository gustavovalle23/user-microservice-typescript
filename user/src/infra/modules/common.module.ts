import { Module } from '@nestjs/common';
import { GraphqlModule } from './graphql.module';
import { MongoModule } from './mongo.module';
import { ConfigModule } from './config.module';

@Module({
  imports: [ConfigModule, GraphqlModule, MongoModule],
  exports: [ConfigModule, GraphqlModule, MongoModule],
})
export class CommonModule {}
