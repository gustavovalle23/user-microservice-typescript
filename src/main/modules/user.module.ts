import { Module } from '@nestjs/common';

import { UserResolver } from '@/main/resolvers';
import { MongooseModule } from '@nestjs/mongoose';
import { User, UserSchema } from '@/infra/repositories/models';
import { MongooseUserRepository } from '@/infra/repositories';
import { JwtGateway } from '@/infra/gateways';
import { JWT_SERVICE } from '@/domain/contracts/gateways';
import { FindUserUseCase } from '@/use-cases/find';
import { FindAllUsersUseCase } from '@/use-cases/list';
import { CreateUserUseCase } from '@/use-cases/create';
import { UserRepository } from '@/domain/contracts/repo';
import { UserService } from '@/main/controller';

@Module({
  imports: [
    MongooseModule.forFeature([
      {
        name: User.name,
        schema: UserSchema,
      },
    ]),
  ],
  controllers: [UserService],
  providers: [
    UserService,
    UserResolver,
    FindUserUseCase.UseCase,
    FindAllUsersUseCase.UseCase,
    CreateUserUseCase.UseCase,

    {
      provide: UserRepository,
      useClass: MongooseUserRepository,
    },
    {
      provide: JWT_SERVICE,
      useClass: JwtGateway,
    },
  ],
})
export class UsersModule {}
