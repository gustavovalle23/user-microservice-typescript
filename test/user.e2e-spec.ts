import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import { print } from 'graphql/language/printer';
import * as request from 'supertest';
import { gql } from 'apollo-server-express';
import { AppModule } from '@/infra/modules';

describe('User (e2e)', () => {
  let app: INestApplication;

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });

  afterEach(async () => {
    await app.close();
  });

  it('Should return a user', async () => {
    await request(app.getHttpServer())
      .post('/graphql')
      .send({
        query: print(gql`
          query FindUserById {
            findUserById(id: "62f5b071f6d8335216f12df2") {
              user {
                id
                name
                isActive
                email
                documentNo
                birthDate
              }
            }
          }
        `),
      })
      .expect(({ body }) => {
        expect(body.errors).toBeUndefined();
        expect(body.data?.findUserById?.user).toMatchSnapshot();
      });
  });

  it('Should return an error when send a invalid user id', async () => {
    await request(app.getHttpServer())
      .post('/graphql')
      .send({
        query: print(gql`
          query FindUserById {
            findUserById(id: "62f5b071f6d8335216f12df1") {
              user {
                id
                name
                isActive
                email
                documentNo
                birthDate
              }
            }
          }
        `),
      })
      .expect(({ body }) => {
        expect(body.errors).toBeDefined();
        expect(body.errors[0]).toMatchObject({
          message: 'User not found with id 62f5b071f6d8335216f12df1',
          extensions: {
            exception: { name: 'UserNotFound' },
          },
        });
      });
  });

  it('Should return all users', async () => {
    await request(app.getHttpServer())
      .post('/graphql')
      .send({
        query: print(gql`
          query AllUsers {
            allUsers {
              users {
                id
                name
                isActive
                email
                documentNo
                birthDate
              }
            }
          }
        `),
      })
      .expect(({ body }) => {
        expect(body.errors).toBeUndefined();
        expect(body.data?.allUsers).toMatchSnapshot();
      });
  });

  it('Should create a user', async () => {
    await request(app.getHttpServer())
      .post('/graphql')
      .send({
        query: print(gql`
          mutation Mutation {
            createUser(
              user: {
                birthDate: "1999-06-26"
                documentNo: "44744675822"
                email: "gustavo.vallerp@hotmail.com"
                isActive: true
                name: "Gusta"
                password: "123456"
              }
            ) {
              user {
                id
                name
                isActive
                email
                documentNo
                birthDate
              }
            }
          }
        `),
      })
      .expect(({ body }) => {
        expect(body.errors).toBeUndefined();
        expect(body.data?.allUsers).toMatchSnapshot();
      });
  });
});
