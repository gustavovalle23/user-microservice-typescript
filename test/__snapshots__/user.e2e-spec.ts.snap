// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User (e2e) Should create a user 1`] = `undefined`;

exports[`User (e2e) Should return a user 1`] = `
Object {
  "birthDate": "1999-06-26T00:00:00.000Z",
  "documentNo": "44744675822",
  "email": "gustavo.vallerp@hotmail.com",
  "id": "62f5b071f6d8335216f12df2",
  "isActive": true,
  "name": "Gusta",
}
`;

exports[`User (e2e) Should return all users 1`] = `
Object {
  "users": Array [],
}
`;
